### ISSUE TITLE
[moe-cpu-native-dtype] Operate on bf16 weights directly in the CPU kernel (no float32 materialization) -- matches upstream FreeToken's design

### ISSUE THREAD

**Issue by aangelinsf** [Performant-Labs/FreeToken-Intel#257]
## Slug
`moe-cpu-native-dtype`

## Why
#251's live validation found hybrid still losing to offload/cpu by ~4x on the 10.2B stand-in even after #248's GIL fix landed and was confirmed working (`job_wait_ms` ~1ms). Root cause: `cpu_moe.cpp`'s kernel (from #250/#252) requires `const float*` input -- every call must materialize a float32 copy of the touched expert weight bank first. On this stand-in (6 experts, ~126MB/expert bf16) that conversion (`submit_ms`, ~17-29ms/layer) costs more than the actual compute it feeds (~12-35ms/layer xpu_half).

Direct comparison against upstream FlashML-org/FreeToken's actual CPU kernel (`python/freetoken/kernel/csrc/cpu_moe/cpu_moe_ext.cpp`, read in full) shows they never hit this problem, because they never made the design choice that causes it: their expert weight banks stay in their native low-precision format (bf16, NVFP4, MXFP4, Q4_0) **permanently** -- never converted to float32 at all. Dequantization happens *inside the innermost SIMD dot-product loop*, fused into the same instruction that reads the weight byte for the multiply:

\`\`\`cpp
// bf16: widened straight into the FMA -- no materialized fp32 copy ever exists
__m512 wf = _mm512_castsi512_ps(_mm512_slli_epi32(_mm512_cvtepu16_epi32(wi), 16));
__m512 xf = _mm512_castsi512_ps(_mm512_slli_epi32(_mm512_cvtepu16_epi32(xi), 16));
acc = _mm512_fmadd_ps(wf, xf, acc);

// or, with native bf16 dot-product hardware support (AVX-512-BF16):
a0 = _mm512_dpbf16_ps(a0, load_bh(w + i), load_bh(x + i));
\`\`\`

Their own file comment states the goal explicitly: *"the C++ W4A16 GEMV reads a row in place ... and dequantizes weights inside the K-loop."* `cpu_executor.py`'s `decode()`/`decode_submit()` confirm the Python side never touches weight tensors at all -- only activations and routing cross the language boundary; the weight bank pointer is resolved once at construction and read in native format from then on. There is no cache because there is nothing to cache: the "conversion" is free, fused into the same bandwidth-bound memory read the GEMV already pays for.

Decision (discussed in #251): redesign our kernel to match this architecture, rather than bolt on a float32 weight cache. A cache would work but keeps the current `const float*` ABI, costs 2x weight memory, and adds cache-invalidation surface upstream's design never needed in the first place.

## Fill in
- New dot-product primitive(s) in `cpu_moe.cpp` operating directly on bf16 weight bytes (`uint16_t`/`bf16_t`), not `float`: a scalar bf16->fp32-widen-and-FMA fallback (matching upstream's `dot_scalar`/`dot_avx512f` shape) and an AVX-512 path doing the same widen-in-register trick #252's existing AVX-512 path already demonstrates the pattern for (just applied to weight bytes instead of pre-converted floats). Runtime-detect and use `AVX512-BF16`'s native `_mm512_dpbf16_ps` when the host CPU supports it (a further, free speedup over the widen-and-FMA fallback), same runtime-dispatch philosophy `avx512_available()` already establishes in this file.
- Change the kernel's `extern "C"` ABI: `gate_up`/`down` become `const uint16_t*` (raw bf16 bit patterns) instead of `const float*`. This is a real interface change -- update `cpu_moe.py`'s ctypes bindings (`_bind_forward`, `_bind_forward_fast`, `_bind_submit` all reference `ctypes.POINTER(ctypes.c_float)` for these params today) and every call site.
- Remove the now-unnecessary `.to("cpu", dtype=torch.float32)` conversions in `cpu_moe.py` (`cpu_moe_forward`, `cpu_moe_forward_fast`, `cpu_moe_submit`, `hybrid_subset_submit`) for `gate_up`/`down` specifically -- these become a straight `.to("cpu")` (keep the model's native bf16), eliminating the O(bank size) conversion pass entirely, matching upstream's zero-materialization design. `x`/activations can stay as they are today (that conversion was never the bottleneck).
- Correctness: output must stay numerically consistent (within the same float32-accumulate tolerance the existing tests use) with the current float32-input kernel -- the accumulate/reduction math is unchanged, only the weight *storage* format changes.
- `hybrid_subset_submit`'s compact-gather path (index_select + remap) stays as-is architecturally; it operates on the bf16 tensors directly now instead of float32-converting first, so its own cost drops too (index_select on bf16 is cheaper than on float32 -- half the bytes moved).

## Test strategy
- Extend the existing `tests/test_kernel_cpu_moe.py` / `tests/test_kernel_cpu_moe_dispatch.py` suites: same synthetic fixtures, but construct `gate_up`/`down` as bf16 tensors and assert the new kernel's output matches the existing pure-Python bf16-aware reference (`_cpu_subset_math`'s original math, or the reference already used in these tests) within float32 tolerance.
- A throughput comparison (not just correctness): per-layer `submit_ms`/`xpu_half_ms` before/after on the #245/#247/#251 live B70 stand-in benchmark -- this issue's whole point is closing the gap #251 measured, so demonstrate the actual reduction, not just that output is still correct.
- Full existing CPU suite must stay green (baseline: 219 passed / 103 skipped, `.venv`; ~675 passed under `.venv-xpu`), zero regressions.

## Accept
- `cpu_moe.cpp`'s hybrid-dispatch entry points operate on native bf16 weight bytes; no `float32`-materialized copy of a full expert weight bank is created anywhere on the hybrid CPU-half path.
- Live B70 re-validation (#251's own benchmark) shows hybrid decode tok/s >= max(offload, cpu) on the 10.2B stand-in -- the epic's (#249) accept bar, still unmet as of this issue's filing.
- No numerical regression vs. the current float32-input kernel.

Depends on: #250, #248 (both already merged; this modifies their code)
Parent epic: #249

Upstream reference: FlashML-org/FreeToken, `python/freetoken/kernel/csrc/cpu_moe/cpu_moe_ext.cpp` (read directly for this issue, not secondhand).

**Comment by aangelinsf** [comment]
Implemented (worktree, not pushed): `worktree-agent-a9f2b9a4fed71315b`, path `/home/andre-angelantoni/Projects/FreeToken-Intel/.claude/worktrees/agent-a9f2b9a4fed71315b`, HEAD `f202066`.

## What was built

Three commits:

1. **`4a99da9`** -- new bf16-native dot-product primitives in `cpu_moe.cpp`:
   - `bf16_to_f32` (exact bit-widen: bf16 is a truncated float32, so `bits << 16` reinterpreted as float32 recovers the value losslessly).
   - `dot_scalar_bf16w` -- scalar widen-and-FMA fallback, correct on any host.
   - `dot_avx512_bf16w` -- the widen-in-register AVX-512F trick from the issue body (`_mm512_cvtepu16_epi32` + `_mm512_slli_epi32(...,16)` + `_mm512_castsi512_ps`, then `_mm512_fmadd_ps`), gated behind `__attribute__((target("avx512f,fma")))` + the existing `avx512_available()` runtime dispatch.
   - A new `moe_forward_core_bf16` core (mirrors `moe_forward_core`'s exact expert-major/top-k-column accumulation order) built on these, with a `force_scalar` flag for deterministic fallback-branch test coverage (mirrors #252's own hook).
   - Native AVX-512-BF16 (`_mm512_dpbf16_ps`) was **not** implemented -- it's explicitly a bonus in the issue, and doesn't actually fit this kernel's shape: it needs *both* dot-product operands in bf16 lanes, but activations (`x`/`act`) deliberately stay float32 per the issue's own scope note, so there's no second bf16 operand to pair it with here.

2. **`8b7b291`** -- ABI + ctypes changes:
   - `freetoken_cpu_moe_submit`'s `gate_up`/`down` params become `const uint16_t*` (raw bf16 bit patterns); `AsyncJob`/`run_job` now route through `moe_forward_core_bf16`.
   - `_bind_submit`'s argtypes updated to `POINTER(c_uint16)`; added `_as_c_bf16_bits_ptr` (a `.view(torch.uint16)` bit-reinterpretation, not a numeric cast -- ctypes/numpy have no bf16 concept).
   - `cpu_moe_submit`: `gate_up`/`down` conversion changed from `.to("cpu", dtype=torch.float32)` to `.to(device="cpu", dtype=torch.bfloat16)` -- the float32 weight-bank materialization on this path is gone. Non-bf16 inputs (e.g. still-float32 tensors) are cast to bf16 here rather than rejected.
   - `hybrid_subset_submit`'s `index_select` gather now runs on bf16 tensors directly (no float32 conversion before or after), so its own cost drops too, per the issue's note.
   - `freetoken_cpu_moe_forward` / `freetoken_cpu_moe_forward_fast` (the plain `cpu` backend) are **left float32-only, unchanged** -- #257's own accept criteria scopes this to the hybrid dispatch path, and the plain backend was never part of the #251 regression.

3. **`f202066`** -- fixed a `NameError` (`_as_c_bf16_bits_ptr` needed its own lazy `import torch`, same pattern every other tensor-touching function in this module already uses) found while running the real test suite, and extended `tests/test_kernel_cpu_moe_dispatch.py`:
   - Fixtures now construct `gate_up`/`down` as `torch.bfloat16`.
   - New bf16-aware Python reference (`.float()` upcast, the same exact widen the kernel performs internally) with a documented tolerance (`rtol=1e-3, atol=1e-4` -- not for bf16 precision loss, which doesn't apply since both sides see the same already-rounded values, but for AVX-512's different-but-equally-valid float32 reduction order vs the scalar loop, matching this repo's existing precedent for that class of comparison).
   - New tests: `force_scalar=True` hook, AVX-512-vs-forced-scalar agreement (this sandbox has AVX-512F, not AVX-512-BF16), the `FREETOKEN_CPU_MOE_FORCE_SCALAR` env var, and non-bf16 dtype inputs getting cast rather than rejected.
   - `tests/test_kernel_cpu_moe.py` (the plain float32 `cpu` backend's tests) is **unchanged** -- its entry points weren't touched.

## Test results

- `.venv-xpu` (the only local venv with `torch` installed), `-k "not xpu"`: **679 passed**, 134 deselected, **1 failed** -- `test_moe_hybrid_profile.py::test_fetch_fraction_none_when_no_profile`, confirmed failing identically on `main` before any of this branch's changes (an env-var/cache-dir test-isolation issue, unrelated to this issue).
- `.venv` (no `torch` installed -- all `torch`-dependent tests, including every `cpu_moe` test, skip): **219 passed / 103 skipped**, unchanged from the stated baseline.

No numerical regressions found; the naive scalar float32 path (`cpu_moe_forward`) and the new bf16-native async path (`cpu_moe_submit`) agree within the documented tolerance across every fixture in both test files.

## Not done / left for the orchestrating session

- Live B70 re-validation (#251's benchmark) and the before/after `submit_ms` throughput comparison this issue's accept bar actually depends on -- no XPU/B70 hardware in this sandbox, as expected.
- Native AVX-512-BF16 fast path -- explicitly optional, and doesn't fit this kernel's mixed-dtype (bf16 weight, float32 activation) shape; noted above.
